<h2 align='center'>NLP Tutorial: Text Representation - Bag Of Words (BOW)</h2>

In [61]:
import pandas as pd
import numpy as np

In [62]:
df = pd.read_csv("spam.csv")
df

,Category,Message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will ü b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...


In [63]:
df.Category.value_counts()

Category
ham     4825
spam     747
Name: count, dtype: int64

Explanation:

    This checks how many messages belong to each category (e.g., "spam" or "ham").

    Category is likely a column indicating whether a message is spam or not.

    value_counts() returns a count of unique values.

In [64]:
df.shape

(5572, 2)

Explanation:

    Returns the dimensions of the DataFrame.

    Output: (number_of_rows, number_of_columns)

In [65]:
df

,Category,Message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will ü b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...


Explanation:

    This again displays the entire DataFrame.

    Probably used to verify what data is loaded after filtering or transformations.

<h3>Train test split</h3>

In [66]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df.Message, df.Category, test_size=0.2, random_state=1)

Explanation:

    This splits the dataset into training and testing sets.

    df.Message = text data (input).

    df.Category = labels (output).

    test_size=0.2 means 20% of the data is used for testing, 80% for training.

    random_state=1 ensures reproducibility of the split.

In [67]:
X_train.shape

(4457,)

Explanation:

    Returns the shape of the training input data.

    Useful to verify the number of training samples.



In [68]:
X_test.shape

(1115,)

Explanation:

    Returns the shape of the testing input data.

In [69]:
X_test

1078                         Yep, by the pretty sculpture
4028        Yes, princess. Are you going to make me moan?
958                            Welp apparently he retired
4642                                              Havent.
4674    I forgot 2 ask ü all smth.. There's a card on ...
                              ...                        
324     That would be great. We'll be at the Guild. Co...
1163    Free entry in 2 a wkly comp to win FA Cup fina...
86      For real when u getting on yo? I only need 2 m...
4214                     I attended but nothing is there.
90      Yeah do! Don‘t stand to close tho- you‘ll catc...
Name: Message, Length: 1115, dtype: object

Explanation:

    Displays the actual test input messages (20% of the dataset).

    Helps to inspect what text data is in the test set.

In [70]:
X_train

1642    Hi , where are you? We're at  and they're not ...
2899          If you r @ home then come down within 5 min
480     When're you guys getting back? G said you were...
3485    Tell my  bad character which u Dnt lik in me. ...
157                           I'm leaving my house now...
                              ...                        
905     We're all getting worried over here, derek and...
5192    Oh oh... Den muz change plan liao... Go back h...
3980    CERI U REBEL! SWEET DREAMZ ME LITTLE BUDDY!! C...
235     Text & meet someone sexy today. U can find a d...
5157                              K k:) sms chat with me.
Name: Message, Length: 4457, dtype: object

Explanation:

    Displays the training input messages.

    Useful for validation or understanding sample text messages.

In [71]:
type(y_train)

pandas.core.series.Series

Explanation:

    Checks the Python type of y_train.

    Expected to be a pandas.Series containing the labels (e.g., "spam" or "ham").

In [72]:
X_train.values

array(["Hi , where are you? We're at  and they're not keen to go out i kind of am but feel i shouldn't so can we go out tomo, don't mind do you?",
       'If you r @ home then come down within 5 min',
       "When're you guys getting back? G said you were thinking about not staying for mcr",
       ...,
       'CERI U REBEL! SWEET DREAMZ ME LITTLE BUDDY!! C YA 2MORO! WHO NEEDS BLOKES',
       'Text & meet someone sexy today. U can find a date or even flirt its up to U. Join 4 just 10p. REPLY with NAME & AGE eg Sam 25. 18 -msg recd@thirtyeight pence',
       'K k:) sms chat with me.'], dtype=object)

Explanation:

    Converts the training message Series into a NumPy array.

    .values gives raw data from a pandas object.

In [73]:
X_train.values.shape

(4457,)

Explanation:

    Shows the shape of the NumPy array that holds the training messages.

In [74]:
type(X_train.values)

numpy.ndarray

Explanation:

    Confirms that X_train.values is indeed a numpy.ndarray.

<h3>Create bag of words representation using CountVectorizer</h3>

In [75]:
from sklearn.feature_extraction.text import CountVectorizer

v = CountVectorizer()

Explanation:

    Imports CountVectorizer, which converts text into a bag-of-words matrix.

    v = CountVectorizer() creates the vectorizer object that will be used to extract features from the text data.

In [76]:
emails = [
            'Upto 20% discount reward! Upto',
            'Hey Aksadur, discount Upto'
         ]

emails_cv = v.fit_transform(emails)
emails_cv

<2x6 sparse matrix of type '<class 'numpy.int64'>'
	with 8 stored elements in Compressed Sparse Row format>

Explanation:

    Creates a list emails with two example email strings.

    Applies CountVectorizer (v) to fit and transform this list into a sparse matrix emails_cv, which contains word counts (bag-of-words).

    emails_cv is a sparse matrix representation of word frequencies for each email.

In [77]:
emails_cv.toarray()

array([[1, 0, 1, 0, 1, 2],
       [0, 1, 1, 1, 0, 1]])

| Email Index | 20 (0) | aksadur (1) | discount (2) | hey (3) | reward (4) | upto (5) |
| ----------- | ------ | ----------- | ------------ | ------- | ---------- | -------- |
| **0**       | 1      | 0           | 1            | 0       | 1          | 2        |
| **1**       | 0      | 1           | 1            | 1       | 0          | 1        |


This is a Bag of Words (BoW) matrix generated by CountVectorizer, where:

    Each row = an email.

    Each column = a word from the vocabulary.

    Each value = the count of that word in the respective email.


✅ Row 0: 'Upto 20% discount reward! Upto'
| Word     | Count |
| -------- | ----- |
| 20       | 1     |
| aksadur  | 0     |
| discount | 1     |
| hey      | 0     |
| reward   | 1     |
| upto     | 2     |
→ Row in array: [1, 0, 1, 0, 1, 2]

In [78]:
v.get_feature_names_out()

array(['20', 'aksadur', 'discount', 'hey', 'reward', 'upto'], dtype=object)

Explanation:

    Displays the vocabulary (word list) extracted from the emails.

    The order of these words corresponds to the columns in the array from the previous cell.

In [79]:
pd.DataFrame(emails_cv.toarray(), columns=v.get_feature_names_out())

,20,aksadur,discount,hey,reward,upto
0,1,0,1,0,1,2
1,0,1,1,1,0,1


Explanation:

    Wraps the array into a DataFrame with proper column names (words).

    Makes it easier to read and interpret the bag-of-words matrix.

In [80]:
v = CountVectorizer()

X_train_cv = v.fit_transform(X_train)
X_train_cv

<4457x7711 sparse matrix of type '<class 'numpy.int64'>'
	with 58978 stored elements in Compressed Sparse Row format>

Explanation:

    Re-initializes the CountVectorizer.

    Fits it on the actual training data (X_train) and transforms it into a bag-of-words matrix (X_train_cv).

In [81]:
X_train_cv.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

Explanation:

    Converts the sparse X_train_cv matrix to a NumPy array to view actual word counts.

In [82]:
X_train_cv.shape

(4457, 7711)

Explanation:

    Shows the dimensions of the transformed training data.

    Shape = (number_of_emails, number_of_unique_words)



In [83]:
X_train_cv[0].toarray()

array([[0, 0, 0, ..., 0, 0, 0]])

Explanation:

    Displays the word count vector for the first training email.

    .toarray() makes it human-readable.

In [84]:
v.get_feature_names_out()

array(['00', '000', '008704050406', ..., 'zyada', 'èn', '〨ud'],
      dtype=object)

Explanation:

    Lists all the vocabulary (features/words) learned from the training set (X_train).

In [85]:
pd.DataFrame(X_train_cv.toarray(), columns=v.get_feature_names_out())

,00,000,008704050406,0121,01223585236,01223585334,0125698789,02,0207,02072069400,...,zed,zeros,zhong,zindgi,zoe,zoom,zouk,zyada,èn,〨ud
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4452,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4453,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4454,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4455,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Explanation:

    Constructs a readable DataFrame showing each training email's word counts for each word in the vocabulary.

In [86]:
v.vocabulary_

{'hi': 3423,
 'where': 7442,
 'are': 1051,
 'you': 7674,
 'we': 7370,
 're': 5564,
 'at': 1119,
 'and': 963,
 'they': 6794,
 'not': 4813,
 'keen': 3906,
 'to': 6889,
 'go': 3153,
 'out': 4985,
 'kind': 3940,
 'of': 4872,
 'am': 937,
 'but': 1578,
 'feel': 2795,
 'shouldn': 6077,
 'so': 6230,
 'can': 1632,
 'tomo': 6910,
 'don': 2408,
 'mind': 4474,
 'do': 2377,
 'if': 3593,
 'home': 3469,
 'then': 6782,
 'come': 1908,
 'down': 2433,
 'within': 7511,
 'min': 4471,
 'when': 7438,
 'guys': 3278,
 'getting': 3116,
 'back': 1196,
 'said': 5845,
 'were': 7419,
 'thinking': 6804,
 'about': 759,
 'staying': 6408,
 'for': 2927,
 'mcr': 4392,
 'tell': 6710,
 'my': 4650,
 'bad': 1199,
 'character': 1738,
 'which': 7448,
 'dnt': 2376,
 'lik': 4108,
 'in': 3633,
 'me': 4393,
 'll': 4156,
 'try': 7007,
 'change': 1730,
 'lt': 4243,
 'gt': 3254,
 'add': 813,
 'tat': 6671,
 'new': 4743,
 'year': 7650,
 'resolution': 5708,
 'waiting': 7308,
 'ur': 7147,
 'reply': 5688,
 'be': 1270,
 'frank': 2963,
 'go

<h3>Train the naive bayes model</h3>

In [87]:
from sklearn.naive_bayes import MultinomialNB
model = MultinomialNB()
model.fit(X_train_cv, y_train)

MultinomialNB()

Explanation
📌 from sklearn.naive_bayes import MultinomialNB

    Imports the Multinomial Naive Bayes classifier from Scikit-learn.

    This algorithm is ideal for text classification problems where features are discrete (like word counts or frequencies).

📌 model = MultinomialNB()

    Initializes a Naive Bayes model instance.

    This model will learn to differentiate between classes (spam vs ham) based on word frequencies.

📌 model.fit(X_train_cv, y_train)

    Trains the model on the training data:

        X_train_cv = bag-of-words representation of training messages.

        y_train = labels (e.g., spam or ham).

    After this step, the model has learned the probability distributions of words given each class.



In [88]:
X_test

1078                         Yep, by the pretty sculpture
4028        Yes, princess. Are you going to make me moan?
958                            Welp apparently he retired
4642                                              Havent.
4674    I forgot 2 ask ü all smth.. There's a card on ...
                              ...                        
324     That would be great. We'll be at the Guild. Co...
1163    Free entry in 2 a wkly comp to win FA Cup fina...
86      For real when u getting on yo? I only need 2 m...
4214                     I attended but nothing is there.
90      Yeah do! Don‘t stand to close tho- you‘ll catc...
Name: Message, Length: 1115, dtype: object

<h3>Evaluate Performance</h3>

In [89]:
X_test_cv = v.transform(X_test)

Explanation

    X_test contains the test email messages (raw text).

    v is the fitted CountVectorizer (trained on X_train), which already knows the vocabulary.

🔧 What this line does:

    Transforms the test messages into the same Bag-of-Words format as the training data (X_train_cv).

    It does not call .fit() again, which is important because:

        .fit() would learn a new vocabulary, possibly different from the training set.

        .transform() uses the same word index mapping learned during training, ensuring consistency.

In [90]:
X_test_cv.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

Explanation

    X_test_cv is a sparse matrix (efficient memory-saving structure) representing the Bag-of-Words vector of the test emails.

    Calling .toarray() converts it into a dense NumPy array, which:

        Makes it human-readable

        Shows the actual word counts for each test message

        Uses the same vocabulary and column order as the training data (X_train_cv)

📊 Example Output (Hypothetical)
| Word1 | Word2 | ... | "discount" | "reward" | "upto" |
| ----- | ----- | --- | ---------- | -------- | ------ |
| 0     | 1     | ... | 1          | 0        | 2      |
| 1     | 0     | ... | 0          | 1        | 0      |


Each row = one test email
Each column = one word in the vocabulary
Each value = count of that word in the message

In [91]:
X_test_cv.toarray().shape

(1115, 7711)

🧠 Explanation

This returns the shape (dimensions) of the dense array created from your test data's bag-of-words representation.
📊 What the Output Means

The result will be something like:

(n_samples, n_features)

Where:

    n_samples = number of test emails (e.g., 1113 if you did 80/20 train-test split on ~5572 total rows)

    n_features = number of unique words (vocabulary size) learned during vectorization on the training data

✅ Example

(1115, 7456)

    1115 test messages

    7456 unique words (features) learned from training data

In [92]:
model.score(X_test_cv, y_test)

0.989237668161435

Explanation

    This line evaluates the accuracy of your trained MultinomialNB model on the test dataset.

    Specifically, it returns the mean accuracy:
    Accuracy=Number of correct predictionsTotal number of predictions
    Accuracy=Total number of predictionsNumber of correct predictions​

📊 Internally Equivalent To

from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test_cv)
accuracy_score(y_test, y_pred)

But model.score(...) is a shortcut that:

    Calls .predict() internally

    Compares predictions to y_test

    Returns the result as a decimal (e.g., 0.976 for 97.6% accuracy)

✅ Example Output

0.9825918762088972

Means the model correctly classified ~98.26% of test messages as either spam or ham.

In [93]:
from sklearn.metrics import classification_report

y_pred = model.predict(X_test_cv)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         ham       0.99      1.00      0.99       968
        spam       0.98      0.94      0.96       147

    accuracy                           0.99      1115
   macro avg       0.98      0.97      0.98      1115
weighted avg       0.99      0.99      0.99      1115



🧠 Explanation

This block is used for detailed model evaluation beyond just accuracy.
✅ Step-by-step:

    model.predict(X_test_cv)

        Uses the trained model to make predictions on the test data.

    classification_report(y_test, y_pred)

        Compares actual vs. predicted labels.

        Returns precision, recall, and F1-score for each class (spam and ham).

📊 Sample Output (Example)

              precision    recall  f1-score   support

         ham       0.99      1.00      1.00      965
        spam       0.98      0.92      0.95      150

    accuracy                           0.99     1115
   macro avg       0.98      0.96      0.97     1115
weighted avg       0.99      0.99      0.99     1115

🔍 What These Metrics Mean
Metric	Meaning
Precision	Of all predicted spam, how many were actually spam?
Recall	Of all actual spam messages, how many did the model correctly catch?
F1-score	Harmonic mean of precision and recall (overall effectiveness).
Support	Actual number of samples per class in the test set.
✅ Interpretation

If you're building a spam filter:

    High recall on spam is crucial (you want to catch all spam).

    But also balance with precision (don’t misclassify important emails as spam).

In [94]:
emails = [
    'Upto 20% discount on parking, exclusive offer just for you. Dont miss this reward!',
    'Hey mohan, can we get together to watch footbal game tomorrow?'
]

In [95]:
emails_count = v.transform(emails)
emails_count.toarray().shape

(2, 7711)

Explanation

    v.transform(emails):

        Converts the new emails into the same Bag-of-Words vector format using the trained vocabulary from the training data.

        This ensures your model can understand the format.

    .toarray():

        Converts the sparse matrix into a full NumPy array so you can inspect the actual values.

    .shape:

        Returns the shape (rows, columns) of the transformed data.

📊 Output

Since you provided 2 emails, the output will be:

(2, N)

Where N is the number of unique words (features) in your training data (e.g., 7456).
✅ Interpretation

    2 rows → 2 email messages.

    N columns → Number of unique words the model was trained on.

    You can now safely use this emails_count to run:

In [96]:
emails_count.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

🧠 Explanation

This command converts the sparse matrix of the new email data (emails_count) into a dense NumPy array, showing the actual word counts from each email according to the trained vocabulary.

Each value in the array represents how many times a word from the training vocabulary appears in the corresponding new email.
📊 Output Format

You'll get a 2D array like:
array([
  [0, 1, 0, 0, 2, ..., 1],  # for email 1
  [1, 0, 0, 1, 0, ..., 0]   # for email 2
])
    Rows = your 2 emails

    Columns = words from training vocabulary

    Values = count of each word in each email

    Note: Most of the values will be 0 (since most words from the full vocabulary won’t appear in each short email).


In [97]:
#emails_count.toarray().shape
model.predict(emails_count.toarray())

array(['spam', 'ham'], dtype='<U4')

🧠 Explanation

    This line predicts the class labels (e.g., 'spam' or 'ham') for the new emails.

    You are passing emails_count.toarray() as input, which is the Bag-of-Words matrix of your emails, converted to a dense array.

    The model uses the patterns it learned during training to classify the emails.


Which means:

    Email 1 was classified as spam

    Email 2 was classified as ham
